torch_geometric/data/iterable_dataset.py

```diff
@@ -0,0 +1,170 @@
+import os.path as osp
+import re
+import sys
+import warnings
+from collections.abc import Sequence
+from typing import Any, Callable, Iterator, List, Optional, Tuple, Union
+
+import torch.utils.data
+
+from torch_geometric.data import Data
+from torch_geometric.data.makedirs import makedirs
+
+# IndexType = Union[slice, Tensor, np.ndarray, Sequence]
+
+
+class IterableDataset(torch.utils.data.IterableDataset):
+    r"""Dataset base class for creating graph datasets which are loaded as
+    an iterator.
+
+    Args:
+        root (string, optional): Root directory where the dataset should be
+            saved. (optional: :obj:`None`)
+        transform (callable, optional): A function/transform that takes in an
+            :obj:`torch_geometric.data.Data` object and returns a transformed
+            version. The data object will be transformed before every access.
+            (default: :obj:`None`)
+        pre_transform (callable, optional): A function/transform that takes in
+            an :obj:`torch_geometric.data.Data` object and returns a
+            transformed version. The data object will be transformed before
+            being saved to disk. (default: :obj:`None`)
+        pre_filter (callable, optional): A function that takes in an
+            :obj:`torch_geometric.data.Data` object and returns a boolean
+            value, indicating whether the data object should be included in the
+            final dataset. (default: :obj:`None`)
+    """
+    @property
+    def raw_file_names(self) -> Union[str, List[str], Tuple]:
+        r"""The name of the files in the :obj:`self.raw_dir` folder that must
+        be present in order to skip downloading."""
+        raise NotImplementedError
+
+    @property
+    def processed_file_names(self) -> Union[str, List[str], Tuple]:
+        r"""The name of the files in the :obj:`self.processed_dir` folder that
+        must be present in order to skip processing."""
+        raise NotImplementedError
+
+    def download(self):
```

Comment:
I'm personally finding it a bit confusing why we want to add all these unimplemented functions to `IterableDataset`.

I feel it may be more meaningful to either 
- create a more concrete version of this, which takes a list of file paths and processes them as required
- remove the file/remote storage parts of this for now

WDTY?

Reply:
First off, I'm happy to implement whatever the maintainers find to be the most meaningful functionality here. Let me provide some more background on this PR.

I noted that the issue linked in the description of this PR describes having the processing of raw files and transforms as desired functionality from a 'chunked' or grouped data loader.

I use this contributed base class to implement a loader which processes a raw data set, save it in groups of samples (of heterodata graphs) according to the number of samples I'd like to store in a group, then loads the groups into memory several at a time, asynchronously. This allows us to tune the amount memory that our dataset uses for training with large amounts of data that would be prohibitive to load at once, while preventing performance issues associated with loading samples one-by-one from disk and the overhead of having very many small files. The full implementation is something that I can contribute. An idea is to provide this as an example implementation of the IterableDataset in documentation. I would like to know how best to 

I think there is a reason to retain the IterableDataset as a base class - other applications of a Dataset as an Iterable are  loading data from a database, from a stream, or from other sources which require code to execute at the time that a sample is requested by the training code.

Would be great to know what you think about this and how best to contribute.

Reply:
With some of the updates, it's more clear to me. I like the idea of adding such a base class, but in the earlier version of this, the reimplemented methods didn't look to be used by any of the base class-provided functionality.

Reply:
Sorry, I didn't mean to suggest that the idea of having a base class for this was bad, just that I couldn't quite follow what the user was expected to implement to make the `iter` work. It looks like this could be provided in the base class, as (given the current implementation) you do need to work with files?

As it is now, we could subclass `Dataset` and just add an `__iter__`.

I don't have a strong opinion here, the design just isn't clear to me.

Reply:
Yes, one would need to work with files during both processing and data loading in some implementations of an iterable dataset.

Pytorch's IterableDataset is a subclass of Pytorch's Dataset, as well - and those are both data containers in PyTorch Lightning https://pytorch-lightning.readthedocs.io/en/stable/guides/data.html

I think both ways work - either subclassing from Pytorch's IterableDataset or subclassing PyG's Dataset, just let me know. It seemed the most straightforward to inherit from PyTorch and parallel the PyTorch Lightning data container structure in PyG when I implemented a chunked dataloader with this.

Reply:
Well I'm okay either way. Leave it to others to follow up. Thanks for explaining your thoughts.